stm32h7rs xspi  Node with domain clock for peripheral clock configuration

FRASTM opened the pull request:
This PR is adding the xspi nodes to the stm32h7rs serie with domain clock configuration for the peripheral clock  and the XSPIMgr clock
The XSPI peripheral is clocked by the STM32_SRC_PLL2_S and the PLL2 is enabled in the board device tree
```
			clock-names = "xspix", "xspi-ker", "xspi-mgr";
			clocks = <&rcc STM32_CLOCK(AHB5, 5U)>,
				<&rcc STM32_SRC_PLL2_S XSPI2_SEL(1)>,
				<&rcc STM32_CLOCK(AHB5, 14U)>;
```

FRASTM committed:
boards: arm: stm32h7s78 disco kit supporting xspi instance

Add the XSPI 2 which is an octoSPI connection to a octo NOR flash
mx66uw1g45 on the stm32h7s78_dk disco kit

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: flash: stm32 xspi driver supports clock domain config

Add the clock domain configuration for the xspi nodes
Where the DTS defines main clock and peripheral clock sel
plus a XSPIM clock

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: arm: stm32h7rs mcu with external memory area

Defines the corresponding memory area of the MPU for
the external NOR xspi node.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: flash: stm32 xspi drivers supporting the stm32h7r/s mcu

Add the support of the stm32h7rs serie to the
drivers/flash/flash_stm32_xspi driver.
The stm32h7rs has no delayblock

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: st: stm32h7s3l8 nucleo board supporting xspi instance

Add the XSPI 2 which is an octoSPI connection to a octo NOR flash
MX25UW25645GXDI00 (256 Mbits, 1.8 V, 200 MHz, DTR, rww)
on the nucleo_h7s3l8 board.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: clock control: stm32H7RS has a PLL2 & 3 or HCLK5 output

Add the definitions of the PLL2 and PLL3 outputs for the stm32H7RS mcus
and the HCLK 5 which is clock source for the XSPI instance.
and other HCLKn for other peripherals.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Can also select the HCLK 5 to be the domain clock of the XSPI peripheral:
```
		xspi2: spi@5200a000 {
			compatible = "st,stm32-xspi";
			reg = <0x5200a000 0x1000>;
			interrupts = <106 0>;
			clock-names = "xspix", "xspi-ker", "xspi-mgr";
			clocks = <&rcc STM32_CLOCK(AHB5, 12U)>,
				<&rcc STM32_SRC_HCLK5 XSPI2_SEL(0)>,
				<&rcc STM32_CLOCK(AHB5, 14U)>;
			#address-cells = <1>;
			#size-cells = <0>;
			status = "disabled";
		};
```

FRASTM commented:
rebase on d453fb94075c5d107f521f8a1ebaf85f3ead32d7

erwango commented on drivers/flash/flash_stm32_xspi.c:
```diff
@@ -2124,17 +2134,17 @@ static int flash_stm32_xspi_init(const struct device *dev)
 #endif /* XSPI_DCR1_DLYBYP */
 
+#if defined(CONFIG_SOC_SERIES_STM32H7RSX)
+	dev_data->hxspi.Init.MaxTran = 0;
```
MaxTran seems available on few other series as well. Is this actually required or already 0 initialized ?

FRASTM replied:
could be replaced by OCTOSPI_DCR3_MAXTRAN or XSPI_DCR3_MAXTRAN register field

erwango commented on drivers/flash/flash_stm32_xspi.c:
```diff
@@ -2124,17 +2134,17 @@ static int flash_stm32_xspi_init(const struct device *dev)
 
+#if defined(CONFIG_SOC_SERIES_STM32H7RSX)
+	dev_data->hxspi.Init.MaxTran = 0;
+	dev_data->hxspi.Init.MemorySelect = HAL_XSPI_CSSEL_NCS1;
```
Already implemented through:
```
#if defined(HAL_XSPIM_IOPORT_1) || defined(HAL_XSPIM_IOPORT_2)
			.MemorySelect = ((DT_INST_PROP(0, ncs_line) == 1)
					? HAL_XSPI_CSSEL_NCS1
					: HAL_XSPI_CSSEL_NCS2),
#endif
```

FRASTM replied:
removed

erwango commented on drivers/flash/flash_stm32_xspi.c:
```diff
@@ -2124,17 +2134,17 @@ static int flash_stm32_xspi_init(const struct device *dev)
 #if defined(XSPI_DCR1_DLYBYP)
 #if STM32_XSPI_DLYB_BYPASSED
 	dev_data->hxspi.Init.DelayBlockBypass = HAL_XSPI_DELAY_BLOCK_BYPASS;
-#else
-	dev_data->hxspi.Init.DelayBlockBypass = HAL_XSPI_DELAY_BLOCK_ON;
```
Why removing it ? Could be valid for other series ?

FRASTM replied:
restored, better set in the `flash_stm32_xspi_dev_data` structure init

erwango commented on boards/st/stm32h7s78_dk/stm32h7s78_dk.dts:
```diff
@@ -173,6 +174,64 @@
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			slot0_partition: partition@0 {
+				label = "image-0";
+				reg = <0x00000000 DT_SIZE_K(512)>;
+			};
+			slot1_partition: partition@80000 {
+				label = "image-1";
+				reg = <0x0080000 DT_SIZE_K(512)>;
+			};
+			scratch_partition: partition@100000 {
+				label = "image-scratch";
+				reg = <0x00100000 DT_SIZE_K(64)>;
+			};
+			storage_partition: partition@110000 {
+				label = "storage";
+				reg = <0x00110000 DT_SIZE_K(64)>;
+			};
```
If 128MB available, maybe we can use wider slots ?

etienne-lms replied:
`CONFIG_BOOT_MAX_IMG_SECTORS` relates the max number of sectors used for each images MCUBOOT handle, so here it relates only to `slot0_partition` and `slot1_partition`. IIFC the storage area is private to Zephyr, not used/considered by MCUBOOT.

erwango reviewed (changes requested):
Some potential common adapattion with N6 similar IP can be used.

gautierg-st commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -455,6 +461,10 @@ static int stm32_clock_control_get_subsys_rate(const struct device *clock,
 	case STM32_CLOCK_BUS_AHB4:
+#if defined(CONFIG_SOC_SERIES_STM32H7RSX)
+	/* HCLK5 i a possible source clock for some peripherals */
+	case STM32_SRC_HCLK5:
```
STM32_SRC_HCLK1/2/3/4 (added in the previous commit) and STM32_CLOCK_BUS_AHB5 should also be added here.

FRASTM replied:
moved to the other commit

gautierg-st replied:
STM32_SRC_HCLK1/2/3/4 and STM32_CLOCK_BUS_AHB5 cases are still missing in this function.

gautierg-st reviewed (dismissed):
One small change, otherwise LGTM

jfischer-no commented on samples/subsys/usb/dfu/sample.yaml:
```diff
@@ -20,6 +20,7 @@ common:
     - mimxrt1180_evk/mimxrt1189/cm7
     - mimxrt700_evk/mimxrt798s/cm33_cpu0
     - nucleo_f207zg
+    - nucleo_h7s3l8
```
Why is that necessary? I do no see test feature `usb_device` in https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/st/nucleo_h7s3l8/nucleo_h7s3l8.yaml

FRASTM replied:
correct, no need to exclude the nucleo_h7s3l8 at this time

jfischer-no commented on samples/subsys/usb/dfu/sample.yaml:
```diff
@@ -29,6 +30,7 @@ common:
     - frdm_mcxa156
     - lpcxpresso55s69/lpc55s69/cpu0
     - stm32l562e_dk/stm32l562xx/ns
+    - stm32h7s78_dk
```
Please remove test feature usb_device from https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/st/stm32h7s78_dk/stm32h7s78_dk.yaml

FRASTM replied:
But it seems, the stm32h7s78_dk does have a usb_device

jfischer-no replied:
Yes, that is why you should remove the test feature usb_device instead of adding it to the exclude list.

jfischer-no replied:
Because there are several boards in this list and it is getting bigger and bigger, and because there is **no test feature for the flash/flash-partitions/flash-foo-bar**, and because not every board needs `usb_device` test feature, and because I am the maintainer of this area and the noise caused by this changes really annoys me and always breaks someone else's patches just for nothing.

FRASTM replied:
OK removing from the `usb_device` entry of the stm32h7s78_dk.yaml

erwango commented on drivers/flash/flash_stm32_xspi.c:
```diff
@@ -2163,7 +2164,10 @@ static int flash_stm32_xspi_init(const struct device *dev)
 #endif /* XSPIM */
 
-#if defined(DLYB_XSPI1) || defined(DLYB_XSPI2) || defined(DLYB_OCTOSPI1) || defined(DLYB_OCTOSPI2)
+#if !defined(CONFIG_SOC_SERIES_STM32H7RSX) && \
+	(defined(DLYB_XSPI1) || defined(DLYB_XSPI2) || \
+	 defined(DLYB_OCTOSPI1) || defined(DLYB_OCTOSPI2))
```
This PR (https://github.com/zephyrproject-rtos/hal_stm32/pull/273/files#diff-c33769f659d0c0eb2dff17944f2dec9129a18177e106a47f1a553f852cc8d9f6) is removing definition of `DLYB_XSPI1` and `DLYB_XSPI1` for H7RS, so this line can be left untouched

FRASTM replied:
done

erwango commented on drivers/flash/flash_stm32_xspi.c:
```diff
@@ -2444,9 +2448,19 @@ static struct flash_stm32_xspi_data flash_stm32_xspi_dev_data = {
 			.FreeRunningClock = HAL_XSPI_FREERUNCLK_DISABLE,
-#if defined(OCTOSPI_DCR4_REFRESH)
+#if defined(OCTOSPI_DCR1_DLYBYP) || defined(XSPI_DCR1_DLYBYP)
+#if STM32_XSPI_DLYB_BYPASSED
+			.DelayBlockBypass = HAL_XSPI_DELAY_BLOCK_BYPASS;
+#else
+			.DelayBlockBypass = HAL_XSPI_DELAY_BLOCK_ON;
+#endif /* STM32_XSPI_DLYB_BYPASSED */
```
Use of `;` isn't correct.

Also let's align syntax on other properties above
```suggestion
			.DelayBlockBypass = (DT_PROP(STM32_XSPI_NODE, dlyb_bypass)
					? HAL_XSPI_DELAY_BLOCK_BYPASS
					: HAL_XSPI_DELAY_BLOCK_ON),
```
And `STM32_XSPI_DLYB_BYPASSED` definition can be removed.

FRASTM commented on drivers/flash/flash_stm32_xspi.c:
```diff
@@ -2163,7 +2174,10 @@ static int flash_stm32_xspi_init(const struct device *dev)
-#if defined(DLYB_XSPI1) || defined(DLYB_XSPI2) || defined(DLYB_OCTOSPI1) || defined(DLYB_OCTOSPI2)
+#if !defined(CONFIG_SOC_SERIES_STM32H7RSX) && \
+	(defined(DLYB_XSPI1) || defined(DLYB_XSPI2) || \
+	 defined(DLYB_OCTOSPI1) || defined(DLYB_OCTOSPI2))
```
changed since the hal_stm32 is updated

erwango commented:
> This PR is waiting for the [zephyrproject-rtos/hal_stm32#273](https://github.com/zephyrproject-rtos/hal_stm32/pull/273) to be merged and west.yml updated (--> CI fails)

PR has been merged. You can rebase.

FRASTM commented:
modifying the DTS to comply with PR https://github.com/zephyrproject-rtos/zephyr/pull/88646

FRASTM commented:
rebase on 76e1fc7713a4a3f2b50c497afddec0364a34c10b

FRASTM committed:
soc: st: stm32 Kconfig to retrieve the external Flash Base address

This commit is retrieving the config FLASH_BASE_ADDRESS
from the XSPI node of the stm32 device dtsi <reg> property of the
"st,stm32-xspi" node. For example the CONFIG_FLASH_BASE_ADDRESS
 is 0x90000000 and application is linked for that address.
Size is given by the size property of the "st,stm32-xspi-nor" node.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented on drivers/flash/flash_stm32_xspi.c:
```diff
@@ -2468,9 +2467,17 @@ static struct flash_stm32_xspi_data flash_stm32_xspi_dev_data = {
 #endif
 			.FreeRunningClock = HAL_XSPI_FREERUNCLK_DISABLE,
-#if defined(OCTOSPI_DCR4_REFRESH)
+#if defined(OCTOSPI_DCR1_DLYBYP) || defined(XSPI_DCR1_DLYBYP)
+			.DelayBlockBypass = (DT_PROP(STM32_XSPI_NODE, dlyb_bypass)
+					? HAL_XSPI_DELAY_BLOCK_BYPASS
+					: HAL_XSPI_DELAY_BLOCK_ON),
+#endif /* xXSPI_DCR1_DLYBYP */
```
`XSPI_DCR1_DLYBYP` and `DelayBlockBypass` don't seem to be available for H7RS.
But they are valid for STM32H5, so it can be decided to be kept in another commit

FRASTM replied:
put in commit 5a948ac12eb447e0117148a1cd35b41f65f2879e

erwango reviewed (changes requested):
Otherwise LGTM.

FRASTM committed:
drivers: flash: stm32 xspi flash driver set the DelayBlock if exists

The stm32H7RS serie has no DelayBlock Bypass control in its DCR1 register.
For other stm32 devices with DelayBlock bypass control, set the value
directly in the structure.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on drivers/flash/flash_stm32_xspi.c:
```diff
@@ -2468,9 +2467,17 @@ static struct flash_stm32_xspi_data flash_stm32_xspi_dev_data = {
+#if defined(OCTOSPI_DCR3_MAXTRAN) || defined(XSPI_DCR3_MAXTRAN)
+			.MaxTran = 0,
+#endif /* xSPI_DCR3_MAXTRAN */
+#if defined(OCTOSPI_DCR4_REFRESH) || defined(XSPI_DCR4_REFRESH)
```
keeping only XSPI_DCR3_MAXTRAN , XSPI_DCR4_REFRESH

etienne-lms reviewed (commented):
Nitpicking comments. Otherwise LGTM.

FRASTM commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -460,6 +466,10 @@ static int stm32_clock_control_get_subsys_rate(const struct device *clock,
 	case STM32_CLOCK_BUS_AHB3:
 	case STM32_CLOCK_BUS_AHB4:
+#if defined(CONFIG_SOC_SERIES_STM32H7RSX)
+	/* HCLK5 i a possible source clock for some peripherals */
```
fixed

FRASTM commented on boards/st/stm32h7s78_dk/stm32h7s78_dk.yaml:
```diff
@@ -13,6 +13,6 @@ supported:
   - watchdog
   - entropy
   - adc
-  - usb_device
```
please refer to https://github.com/zephyrproject-rtos/zephyr/pull/88051#discussion_r2056516046

FRASTM commented:
rebase on 88b108d201c696667db123d571562f1c9f81f2ac (fixing typo and conflict)

FRASTM commented:
adding the HCLKn clocks

etienne-lms commented on drivers/flash/flash_stm32_xspi.c:
```diff
@@ -2068,6 +2073,11 @@ static int flash_stm32_xspi_init(const struct device *dev)
 		LOG_ERR("XSPI mode SPI|DUAL|QUAD/DTR is not valid");
 		return -ENOTSUP;
 	}
+#if defined(CONFIG_SOC_SERIES_STM32H7RSX)
```
Nitpicking: may I suggest an empty line above?

etienne-lms commented on boards/st/nucleo_h7s3l8/nucleo_h7s3l8.dts:
```diff
@@ -139,3 +152,62 @@
+			slot0_partition: partition@0 {
+				label = "image-0";
+				reg = <0x00000000 DT_SIZE_K(512)>;
+			};
```
Prefer with an empty line before each subnodes.

etienne-lms commented on boards/st/nucleo_h7s3l8/nucleo_h7s3l8.dts:
```diff
@@ -139,3 +152,62 @@
+			};
+			storage_partition: partition@110000 {
+				label = "storage";
+				reg = <0x00110000 DT_SIZE_K(64)>;
```
Intentionally waste the remaining of the internal flash? 64kB is not much. Maybe increase to 1 or 2 MByte, or even `(DT_SIZE_M(30) - DT_SIZE_K(64))`.

etienne-lms commented on boards/st/stm32h7s78_dk/stm32h7s78_dk.dts:
```diff
@@ -207,6 +207,65 @@
+			slot0_partition: partition@0 {
+				label = "image-0";
+				reg = <0x00000000 DT_SIZE_K(512)>;
+			};
```
Add an empty line between each subnode.

etienne-lms reviewed (approved):
LGTM, with these suggestions/comments addressed of not.

etienne-lms reviewed (changes requested):
Sorry, I'm removing my +1, at leas tthe empty lines should be added in the DTS file (coding style conformance).

etienne-lms commented:
So quick! nevermind, we'll fix the DT later :-)
